src/client/core/page-unload-barrier.js

```diff
@@ -100,4 +101,15 @@ export function wait () {
                     waitingPromiseResolvers.push(resolve);
             });
     });
+
+    // NOTE: sometimes the page is not really unloading after the beforeunload event
+    // appeared (see issues #664, #437). To avoid test hanging we resolve unload
+    // barrier waiting promise in MAX_UNLOADING_TIMEOUT. We can improve this logic when
+    // the https://github.com/DevExpress/testcafe-hammerhead/issues/667 issue is fixed.
```

Comment:
```
    // NOTE: sometimes the page isn't actually unloaded after the beforeunload event
    // fires (see issues #664, #437). To avoid test hanging, we resolve the unload
    // barrier waiting promise in MAX_UNLOADING_TIMEOUT. We can improve this logic when
    // the https://github.com/DevExpress/testcafe-hammerhead/issues/667 issue is fixed.
```